### PR TITLE
Support Net functions

### PR DESCRIPTION
**Pull request by logicoffee**
Hi, @goccy 
I implemented NET functions.
https://cloud.google.com/bigquery/docs/reference/standard-sql/net_functions

I'm not very familiar with Golang, so please let me know if there's something wrong.
Thanks!

### CODE REVIEW

File: internal/function_net.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+		}
+	}
+	return parsed
+}
+
+func public_suffix(host string) string {
+	reg, _ := regexp.Compile(`[^.]\.{2,}[^.]`)
+	if reg.MatchString(host) {
+		return ""
+	}
```
Comment (logicoffee) [inline internal/function_net.go:136]: This corresponds to the following specificaiton.

> This function returns NULL if any of the following is true:
> - The parsed host contains adjacent dots in the middle (not leading or trailing);

https://cloud.google.com/bigquery/docs/reference/standard-sql/net_functions?hl=en#netpublic_suffix

File: internal/function_net.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	return BytesValue(ip), nil
+}
+
+func parse_url(v string) *url.URL {
+	parsed, err := url.Parse(v)
+	if err != nil {
+		return nil
+	}
+	if parsed.Host == "" {
+		parsed, err = url.Parse("//" + strings.TrimSpace(v))
```
Comment (logicoffee) [inline internal/function_net.go:124]: To make the test case `    www.Example.Co.UK    ` to pass.

File: internal/function_net.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	return StringValue(suffix), nil
+}
+
+func NET_REG_DOMAIN(v string) (Value, error) {
+	parsed := parse_url(v)
+	if parsed == nil {
+		return nil, nil
+	}
+	host := parsed.Hostname()
+	split_host := strings.Split(host, ".")
```
Comment (goccy) [inline internal/function_net.go:101]: Please use `splitHost` , `splitSuffix` as a variable name .

File: internal/function_net.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+		parsed, err = url.Parse("//" + strings.TrimSpace(v))
+		if err != nil {
+			return nil
+		}
+	}
+	return parsed
+}
+
+func public_suffix(host string) string {
+	reg, _ := regexp.Compile(`[^.]\.{2,}[^.]`)
```
Comment (goccy) [inline internal/function_net.go:133]: Compile is heavy processing and should be performed outside of the function. 

````go
var (
  publicSuffixMatcher = regexp.MustCompile(`[^.] \. {2,}[^.]]`)
)

func publicSuffix(host string) string {
}
````

File: internal/function_net.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		return ""
+	}
+	split_suffix := strings.Split(suffix, ".")
+	return strings.Join(split_host[len(split_host)-len(split_suffix):], ".")
+}
+
+func parse_ip(v string) []byte {
+	ip, err := netip.ParseAddr(v)
+	if err != nil {
+		return nil
```
Comment (goccy) [inline internal/function_net.go:153]: Is it ok to ignore this error?

File: internal/function_net.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+func public_suffix(host string) string {
+	reg, _ := regexp.Compile(`[^.]\.{2,}[^.]`)
+	if reg.MatchString(host) {
+		return ""
+	}
+	split_host := strings.Split(host, ".")
+	encoded, err := idna.ToASCII(strings.ToLower(host))
+	if err != nil {
+		return ""
```
Comment (goccy) [inline internal/function_net.go:140]: Is it ok to ignore this error?

**Review by goccy — COMMENTED** [review]
@logicoffee Thank you for your great PR !!!
I made some comments. Please check 🙏

File: internal/function_net.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	"golang.org/x/net/publicsuffix"
+	"net"
+	"net/netip"
+	"net/url"
+	"regexp"
+	"strings"
+)
+
+func NET_HOST(v string) (Value, error) {
+	parsed := parse_url(v)
```
[the remark this answers is not in the document]
Comment (logicoffee) [reply internal/function_net.go]: Fixed in https://github.com/goccy/go-zetasqlite/pull/99/commits/5510590da4c440515c4f31a3bb669bb7d254444f

File: internal/function_net.go (hunk 4 above)
Comment (logicoffee) [reply internal/function_net.go]: Fixed in https://github.com/goccy/go-zetasqlite/pull/99/commits/8d0120437c985a79cbe39fe003da30c6ac6658b2

File: internal/function_net.go (hunk 6 above)
Comment (logicoffee) [reply internal/function_net.go]: Fixed in https://github.com/goccy/go-zetasqlite/pull/99/commits/b68098425acba41999f40c821ed0e588556dbdef

The logic is clearer when an error is returned 😃 

In addition, the document doesn't specify that `NET.PUBLIC_SUFFIX` returns NULL when normalization fails.
So, the error shouldn't be suppressed I guess.

**Comment by goccy** [comment]
@logicoffee  Thank you for the fixing !! LGTM

**Pull request merged**